packages/apollo-server-plugin-response-cache/src/ApolloServerPluginResponseCache.ts

```diff
@@ -0,0 +1,337 @@
+  // be anything that can be JSON-stringified.
+  extraCacheKeyData?(
+    requestContext: GraphQLRequestContext<TContext>,
+  ): ValueOrPromise<any>;
+
+  // If this hook is defined and returns false, the plugin will not read
+  // responses from the cache.
+  shouldReadFromCache?(
+    requestContext: GraphQLRequestContext<TContext>,
+  ): ValueOrPromise<boolean>;
+
+  // If this hook is defined and returns false, the plugin will not write the
+  // response to the cache.
+  shouldWriteToCache?(
+    requestContext: GraphQLRequestContext<TContext>,
+  ): ValueOrPromise<boolean>;
+}
+
+enum SessionMode {
+  NoSession,
+  Private,
+  AuthenticatedPublic,
+}
+
+function sha(s: string) {
+  return createHash('sha256')
+    .update(s)
+    .digest('hex');
+}
+
+interface BaseCacheKey {
+  documentText: string;
+  operationName: string | null;
+  variables: { [name: string]: any };
+  extra: any;
+}
+
+interface ContextualCacheKey {
+  sessionMode: SessionMode;
+  sessionId?: string | null;
+}
+
+interface CacheValue {
+  // Note: we only store data responses in the cache, not errors.
+  //
+  // There are two reasons we don't cache errors. The user-level reason is that
+  // we think that in general errors are less cacheable than real results, since
+  // they might indicate something transient like a failure to talk to a
+  // backend. (If you need errors to be cacheable, represent the erroneous
+  // condition explicitly in data instead of out-of-band as an error.) The
+  // implementation reason is that this lets us avoid complexities around
+  // serialization and deserialization of GraphQL errors, and the distinction
+  // between formatted and unformatted errors, etc.
+  data: Record<string, any>;
+  cachePolicy: Required<CacheHint>;
+  cacheTime: number; // epoch millis, used to calculate Age header
+}
+
+type CacheKey = BaseCacheKey & ContextualCacheKey;
+
+function cacheKeyString(key: CacheKey) {
+  return sha(JSON.stringify(key));
+}
+
+function isGraphQLQuery(requestContext: GraphQLRequestContext<any>) {
+  return requestContext.operation!.operation === 'query';
+}
+
+export default function plugin(
+  options: Options = Object.create(null),
+): ApolloServerPlugin {
+  return {
+    requestDidStart(
+      outerRequestContext: GraphQLRequestContext<any>,
+    ): GraphQLRequestListener<any> {
+      const cache = new PrefixingKeyValueCache(
+        options.cache || outerRequestContext.cache!,
+        'fqc:',
+      );
+
+      let sessionId: string | null = null;
+      let baseCacheKey: BaseCacheKey | null = null;
+      let age: number | null = null;
+
+      return {
+        async responseForOperation(
+          requestContext: WithRequired<
+            GraphQLRequestContext<any>,
+            'document' | 'operationName' | 'operation'
+          >,
+        ): Promise<GraphQLResponse | null> {
+          requestContext.metrics!.responseCacheHit = false;
```

Comment:
I suspect we could get rid of the non-null `!` assertion here with the `WithRequired` changes for `metrics` I've suggested previously?